### PR TITLE
[WIP] v4 - Counter offer

### PR DESCRIPTION
**Pull request by idinium96**
I will keep this open until completed.

Update:
- Testing phase (~~manually accept with `!fcounter`~~, not started yet on my bots)
- 20/05/2021: Start testing on my bots ([Fumino](https://backpack.tf/profiles/76561198077052056), [Nezuko](https://backpack.tf/profiles/76561198034345222), [Mai](https://backpack.tf/profiles/76561198053911511), [ZeroTwo](https://backpack.tf/profiles/76561198163541169))

## New options.json properties
- `miscSettings.counterOffer.enable` (default is `true`)
- `customMessage.decline.failedToCounter` (default is `"/pre ❌ Ohh nooooes! The offer is no longer available. Reason: The offer has been declined. Counteroffer is not possible because either one of us does not have enough pure."`)

### CODE REVIEW

**Comment by idinium96** [comment]
Current situation after confirming the countered offer:
image

The bot will fail to send a trade summary since we do not have data for `value` and `dict` saved on polldata for the countered offer (new offerID).

**Comment by Preport** [comment]
Screenshot_89
It should work now other than Construct and complete times.

**Comment by idinium96** [comment]
> Screenshot_89
> It should work now other than Construct and complete times.

Thank you. I will also change that `chat` to maybe `offer - countered` - tomorrow

**Comment by idinium96** [comment]
I think making the bot counter an offer to the exact metal/keys is the best.

This is from my test:
image

I set the bot to buy each hat for 0.44 ref, and the counter result is:
image

Which is honestly a bit confusing.

---

The Gladiator bot does the correct way:
image
image

**Comment by Preport** [comment]
It should've removed metal from its side maybe I missed something whilst refactoring.

**Comment by idinium96** [comment]
> It should've removed metal from its side maybe I missed something whilst refactoring.

Oh is it? I thought it was the same.

**Comment by Preport** [comment]
It was probably the same. I think the only thing missing was my mind. It should work now. Did the same test as you.
Screenshot_96
Screenshot_97

**Comment by Preport** [comment]
Or maybe you wanted bot to remove few extra in favor of paying with 1.33 instead of 2 and getting .66 which should be possible by changing 
https://github.com/TF2Autobot/tf2autobot/pull/455/files#diff-7ae71475718c2654e9ec0ed072d18f2241c952c4d06ba4dbfadfb1a2fa9d3dadR846-R847.
these to lines overpay to true
But don't quote me on that :)

**Comment by idinium96** [comment]
Each hat still set to 0.44 ref buying price, 3 of them means 1.33 ref.
I put 1 key on the bot side while the trade partner has only 36 ref metals, and this is the result
image
image

Which is an easy profit for them since the key is currently at 58 ref.
36 ref + 1.33 ref + 0.66 ref = 38 ref

It should fail, right?

---

And also, refined metals are missing in the summary (not recorded in polldata):
image
image

**Comment by idinium96** [comment]
# Critical unknown bug:
The bot suddenly stop processing offer(s) received when sending countered offer (accepting mobile confirmation was not yet executed) and disturbed by `onOfferChanged` event.

### Screenshot 1
Screenshot_2

### Screenshot 2
Screenshot_3

### Logs
[ZeroTwo-2021-05-22.log](https://github.com/TF2Autobot/tf2autobot/files/6526920/ZeroTwo-2021-05-22.log)

**Comment by idinium96** [comment]
image

Not sure how can this be `undefined`

## `dist/classes/Trades.js`
image

## `src/classes/Trades.ts`
image

## Probably because contains craft weapons (not in pricelist, craft weapon as currency enabled)?
image
image

**Comment by idinium96** [comment]
I will fix this

File: src/classes/Trades.ts
Comment (Preport) [inline src/classes/Trades.ts:923]: We count in scraps so shouldn't this be (half a scrap) 0.5 * dataDict[side][sku] (which is the amount of the item)

**Review by Preport — CHANGES_REQUESTED** [review]
I feel like this should be changed

File: src/classes/Trades.ts
Comment (idinium96) [reply src/classes/Trades.ts:923]: Oh right thank you.

Comment (idinium96) [reply src/classes/Trades.ts:923]: `dataDict[side][sku]` shouldn't be undefined.

Comment (Preport) [reply src/classes/Trades.ts:923]: Are you talking about the line `922`?

Comment (idinium96) [reply src/classes/Trades.ts]: No no.

Comment (Preport) [reply src/classes/Trades.ts]: One no was enough 😭.

**Review by Preport — APPROVED** [review]

File: src/classes/Trades.ts
Comment (idinium96) [inline src/classes/Trades.ts:763]: I will not approve until this is changed or implement an option for a custom message 😚

Comment (Preport) [reply src/classes/Trades.ts:763]: Screenshot_131

Comment (idinium96) [reply src/classes/Trades.ts:763]: 🤣🤣🤣

**Review by Preport — APPROVED** [review]
LGTM.

**Review by arik123 — APPROVED** [review]
The code seems fine, I would suggest some more testing before merging into master

**Comment by rennokki** [comment]
Well, things look good, as long as it has testings in place. 🎈

**Review by rennokki — APPROVED** [review]

**Pull request merged**
